### ISSUE TITLE
tf.data.experimental.load does not work on temporary files

### ISSUE THREAD

**Issue by Thf772** [tensorflow/tensorflow#53292]
**System information**
- TensorFlow version (you are using): 2.6.0
- Are you willing to contribute it (Yes/No): No (I don't have the required knowledge)

**Describe the feature and the current behavior/state.**

When loading a dataset with `tf.data.experimental.load`, the produced dataset will always be a lazily-loaded dataset that doesn't read the files until the dataset object is consumed (e.g. by iterating over it with Python code, or using it to train a model). This poses a problem if the path provided to `load` is a temporary path, since depending on when the dataset is used, the files could be gone when the dataset object attempts to read them.

Consider the following load function:
```python
def load_data(source):
    with tempfile.TemporaryDirectory() as tmpdir:
        prepare_files(source, tmpdir.name)
        return tf.data.experimental.load(tmpdir.name)
```
The dataset returned by this function will be unusable since, as soon as the function returns, the `TemporaryDirectory` context manager will exit and destroy the files. (The `prepare_files` function is just there to represent any logic that could be used to get the files to the temp. dir, such as downloading them over network, extracting them from an archive, or decrypting them.)

Aside from the temporary files issue, there could some other cases in which the developpers would want their datasets to be loaded in memory immediately.

**Will this change the current api? How?**

The feature I propose is to allow callers of `tf.data.experimental.load` to specify whether they want a lazily-loaded dataset, or an eagerly-loaded dataset (that reads the files only once, when `load` is called, and stores the data in memory).

The way I see it, this evolution would add a new optional parameter to `tf.data.experimental.load`. Its default value needs to correspond to lazy loading for backwards compatibility.

This would not pose a big issue since 1. it will be backwards-compatible and 2. only the experimetal API will be modified (and the documentation explicetely says the experimental API may evolve at any time).

**Who will benefit with this feature?**

Anyone who wants to separate their dataset loading logic in a similar function, using a temporary directory or another context manager that destroys the files upon exit. This would be the case when:
- using files downloaded from the Internet
- using encrypted files, e.g. if the data is sensitive or private
- using files compressed externally, to have more control over compression
- using `importlib.resources.as_file` in some cases

**Any Other info.**

The fact that the datasets are lazy-loaded is not documented, it would be nice to add this to the current documentation of `tf.data.experimental.load`. Also on the subject of documentation, once (if) the eager load option is added, the documentation should warn the user about the potentially high memory usage when eager-loading large datasets. In my use case this is not a problem, since 1. the datasets are created with `Dataset.from_tensor_slices` before being saved (so they have already been entirely in memory before), 2. I only use one dataset at a time and 3. I make sure it is garbage-collected as soon as possible.

Here is my current placeholder:
```python
def load_data(source):
    tmppath = tempfile.mkdtemp()
    def deleter():
        shutil.rmtree(tmppath)
    atexit.register(tmppath)
    prepare_files(source, tmppath)
    return tf.data.experimental.load(tmppath)
```
Depending on how many times this function is called during the program's lifetime, this can quickly clutter the disk. It would be nice to have a better solution.

Note that, to fit **my** *specific* use case, it would also work for me to have a variant of `save` and `load` that accept a Python file-like object (and save to/load from a single file). The logic I use in my version of `prepare_files` turns a single file into a directory.

**Comment by Thf772** [comment]
By the way, I have already tried calling `cache` on the dataset and iterating over it prior to exiting the `with` block, but this still results in an exception being thrown when the dataset is used outside of the function.

**Comment by Thf772** [comment]
I have re-created the error in the Python console:
```
>>> next(iter(loaded_dataset))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Program Files\Python39\lib\site-packages\tensorflow\python\data\ops\dataset_ops.py", line 411, in __iter__
    return iterator_ops.OwnedIterator(self)
  File "C:\Program Files\Python39\lib\site-packages\tensorflow\python\data\ops\iterator_ops.py", line 696, in __init__
    self._create_iterator(dataset)
  File "C:\Program Files\Python39\lib\site-packages\tensorflow\python\data\ops\iterator_ops.py", line 719, in _create_iterator
    gen_dataset_ops.make_iterator(ds_variant, self._iterator_resource)
  File "C:\Program Files\Python39\lib\site-packages\tensorflow\python\ops\gen_dataset_ops.py", line 3123, in make_iterator
    _ops.raise_from_not_ok_status(e, name)
  File "C:\Program Files\Python39\lib\site-packages\tensorflow\python\framework\ops.py", line 6941, in raise_from_not_ok_status
    six.raise_from(core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.NotFoundError: Could not find metadata file.
         [[{{node LoadDataset/_1}}]] [Op:MakeIterator]
```
`loaded_dataset` is a dataset loaded from a now deleted temporary directory, on which I have done the following in order:
- Call `tf.data.experimental.load` on the temporary path to load the dataset
- Add the `.cache()` operation with no argument (cache to memory)
- Consume the dataset with `for _ in cached_dataset: pass`
- Exit the temporary directory context manager (_after_ the cached dataset has been consumed once)

The error happens as soon as I try to use the dataset outside of the load function.

**Comment by aaudiber** [comment]
I'm surprised using `cache` doesn't avoid the issue. Here is a self-contained repro: https://colab.research.google.com/drive/188Kvhi5bVCtcEP54h3nRix-nrW9-LN3X?usp=sharing. Will need further investigation to see what's going on.

**Comment by shilpakrish2** [comment]
I just sent out a fix in cl/415373976.  Before this change, dataset finalization (which allows options to propagate through datasets) was performed each time an iterator for a dataset was created.  When finalizing a cached LoadDataset, an error was thrown if the underlying file was no longer present. This would occur if the saved file had been in a temporary directory and the temporary directory was now out of scope. This CL fixes the issue by performing finalization only the first time an iterator for a dataset is created. Subsequent iterations use a cached version of the finalized dataset rather than performing the finalization again.
